tests/test_csc.py

```diff
@@ -0,0 +1,136 @@
+
+    async def __aexit__(self, *args):
+        await self.csc.close()
+        await self.server.stop()
+
+
+class CscTestCase(unittest.TestCase):
+    def setUp(self):
+        pass
+
+    def test_standard_state_transitions(self):
+        async def doit():
+            async with Harness(initial_state=salobj.State.STANDBY) as harness:
+                state = await harness.remote.evt_summaryState.next(flush=False, timeout=STD_TIMEOUT)
+                self.assertEqual(state.summaryState, salobj.State.STANDBY)
+
+                await harness.remote.cmd_start.start()
+                self.assertEqual(harness.csc.summary_state, salobj.State.DISABLED)
+                state = await harness.remote.evt_summaryState.next(flush=False, timeout=STD_TIMEOUT)
+                self.assertEqual(state.summaryState, salobj.State.DISABLED)
+
+                await harness.remote.cmd_enable.start()
+                self.assertEqual(harness.csc.summary_state, salobj.State.ENABLED)
+                state = await harness.remote.evt_summaryState.next(flush=False, timeout=STD_TIMEOUT)
+                self.assertEqual(state.summaryState, salobj.State.ENABLED)
+
+                await harness.remote.cmd_disable.start()
+                self.assertEqual(harness.csc.summary_state, salobj.State.DISABLED)
+                state = await harness.remote.evt_summaryState.next(flush=False, timeout=STD_TIMEOUT)
+                self.assertEqual(state.summaryState, salobj.State.DISABLED)
+
+                await harness.remote.cmd_standby.start()
+                self.assertEqual(harness.csc.summary_state, salobj.State.STANDBY)
+                state = await harness.remote.evt_summaryState.next(flush=False, timeout=STD_TIMEOUT)
+                self.assertEqual(state.summaryState, salobj.State.STANDBY)
+
+        asyncio.get_event_loop().run_until_complete(doit())
+
+    def test_apply_position_limits(self):
+        async def doit():
+            async with Harness(initial_state=salobj.State.STANDBY) as harness:
+                await harness.remote.cmd_start.start(timeout=STD_TIMEOUT)
+                await harness.remote.cmd_enable.start(timeout=STD_TIMEOUT)
+                harness.remote.evt_settingsAppliedPositionLimits.flush()
+                await harness.remote.cmd_applyPositionLimits.set_start(timeout=STD_TIMEOUT, xyMax=12, zMin=-6,
+                                                                       zMax=12, uvMax=8, wMin=-6, wMax=6)
+                event = await harness.remote.evt_settingsAppliedPositionLimits.next(flush=False,
+                                                                                    timeout=STD_TIMEOUT)
+                self.assertEqual(12, event.limitXYMax)
+                self.assertEqual(-6, event.limitZMin)
+                self.assertEqual(12, event.limitZMax)
+                self.assertEqual(8, event.limitUVMax)
+                self.assertEqual(-6, event.limitWMin)
+                self.assertEqual(6, event.limitWMax)

```

Comment:
I *strongly* suggest that you use different values for xyMax and zMax.

Please look for this in all your other tests, as well. If you use a different value for each available axis (in this case you specify xyMax, so x and y must have the same value) then that allows you test for the all-to-common error that your code refers to the wrong axis.

(That problem of referring to the wrong axis is especially easy to cause if you use a separate variable for each axis, which is why I suggested you switch to arrays. But it is always a concern. And it can be really hard to spot by reading the code. Unit tests are your best hope of catching it.)

Reply:
I also suggest you use constants to hold these values in your tests. For example:
```
xyMax = 12
set_start(..., xyMax=xyMax)
...
self.assertEqual(xyMax, event.limitXYMax)
```
It makes the code easier to read and allows you to change the values in one place (for example to give each axis a different value!). It also allows you to test multiple values with a loop, should that prove helpful.

Reply:
Does the CSC or low-level controller sanity-check these values?

I hope so.

Please add tests that calling these various functions with "out-of-range" values causes the command to fail, e.g.:
```
with salobj.assertRaisesAckError(...):
    harness.remote.cmd_applyPositionLimits.set_start(timeout=STD_TIMEOUT, xyMax=badXyMax, ...)
```
For completeness I suggest setting each argument invalid in turn (with the others having valid values), to prove that *every* argument is correctly range-tested. A nice way to do that is to start by making a keyword argument dict of valid arguments. Then for each iteration make a copy of the dict and change one value in it to an invalid value.

Reply:
Low level hardware will reject any parameters outside of range or existence.